functorch/_src/make_functional.py

```diff
@@ -56,66 +55,12 @@ def raise_parameter_tying_error():
         "https://github.com/pytorch/functorch/issues/446")
 
 
-def create_names_map(named_params, tied_named_params):
-    """
-    named_params is a dictionary of tensors: {'A': A, 'B': B}
-    tied_named_params is another dictionary of tensors {'A': A, 'B': B, 'B_tied': B}
-    with potentially tied (or 'duplicated') tensors
-
-    This function creates a mapping from the names in named_params to the
-    names in tied_named_params: {'A': ['A'], 'B': ['B', 'B_tied']}.
-    """
-    named_params = {k: v for k, v in named_params}
-    tied_named_params = {k: v for k, v in tied_named_params}
-
-    tensors_dict_keys = set(named_params.keys())
-    tied_tensors_dict_keys = set(tied_named_params.keys())
-    assert tensors_dict_keys.issubset(tied_tensors_dict_keys)
-
-    tensor_to_mapping = {}
-    for key, tensor in named_params.items():
-        tensor_to_mapping[tensor] = (key, [])
-    for key, tensor in tied_named_params.items():
-        assert tensor in tensor_to_mapping
-        tensor_to_mapping[tensor][1].append(key.split('.'))
-    result = {key: value for key, value in tensor_to_mapping.values()}
-    return result
-
-
-def _extract_members(mod: nn.Module, _named_members, named_members, subclass):
-    all_named_members = tuple(_named_members(mod, remove_duplicate=False))
-    named_members = tuple(named_members())
-    names_map = create_names_map(named_members, all_named_members)
-
-    # Remove all the members in the model
-    memo = {}
-    for name, p in all_named_members:
-        if p not in memo:
-            memo[p] = subclass(torch.empty_like(p, device='meta'))
-        replacement = memo[p]
-        _set_nested_attr(mod, name.split("."), replacement)
-
-    if len(named_members) == 0:
-        names, params = (), ()
-    else:
-        names, params = zip(*named_members)
-    return params, names, names_map
-
-
-def extract_weights(mod: nn.Module):
-    """
-    This function removes all the Parameters from the model and
-    return them as a tuple as well as their original attribute names.
-    The weights must be re-loaded with `load_weights` before the model
-    can be used again.
-    Note that this function modifies the model in place and after this
-    call, mod.parameters() will be empty.
-    """
-    return _extract_members(mod, _named_parameters, mod.named_parameters, nn.Parameter)
-
-
-def extract_buffers(mod: nn.Module):
-    return _extract_members(mod, _named_buffers, mod.named_buffers, lambda x: x)
+def extract_weights(model):
+    for module_name, m in model.named_modules():
+        for param_name, p in list(m.named_parameters(recurse=False)):
+            delattr(m, param_name)
+            setattr(m, param_name, None)
+            yield (module_name, m, param_name, p)
```

Comment:
The reason we previously used create_names_map was for parameter tying. If someone creates a module that looks like:
```
class Foo(nn.Module):
   def __init__(self):
        super().__init__()
        self.bias = nn.Parameter(torch.randn(3))
        self.linear = nn.Linear(3, 3)
        self.linear.bias = self.bias
```
then `fmodel, params = make_functional(Foo())` returns 2 Tensors (self.linear.weight and self.bias) instead of 3 Tensors. When the user calls `fmodel([w, b], x)`, then b gets loaded to self.bias and self.linear.bias and w gets loaded to self.linear.weight.

Under the new strategy, it seems like `params` would have 3 tensors: [self.bias, self.linear.weight, self.linear.bias]. 

In general I'm not really sure what the interaction between parameter tying and make_functional should be. Thoughts?

Reply:
I see your point. If we want to keep things as they are we could link params to a list of modules and a list of names (instead of a single module and a single name). That will come with a slight overhead though...

It's the kind of design choice where you will always make someone unhappy (there will be someone out there that wants multiple copies of the same param), but it's probably not the majority of users.